### PR TITLE
Coproducts

### PR DESCRIPTION
**Pull request by Fristi**
Allows you to do coproducts in map by adding discrimantor key `__${caseName} -> Map.empty`

### CODE REVIEW

**Comment by Fristi** [comment]
I added `xmap` on BigQueryType. It can be used to support to support `ByteVector` and date time values without the need to pull in those dependencies in your project.

Example given:

```scala
def toInstant(v: String): Instant = timestampParser.parseDateTime(v).toInstant
def fromInstant(i: Instant): String = timestampPrinter.print(i)

implicit val instant = BigQueryMappableType.stringBigQueryMappableType.xmap(toInstant)(fromInstant)
```

This would remove the need to have these libs as dependency and would also apply to datastore and such.

**Comment by Fristi** [comment]
Builds seems to be stuck, I'll have a look later :-)

File: bigquery/src/test/scala/shapeless/datatype/bigquery/BigQueryTypeSpec.scala [hunk 1]
```diff
@@ -18,7 +18,9 @@ class BigQueryTypeSpec extends Properties("BigQueryType") {
   implicit def compareByteArrays(x: Array[Byte], y: Array[Byte]) = java.util.Arrays.equals(x, y)
   implicit def compareIntArrays(x: Array[Int], y: Array[Int]) = java.util.Arrays.equals(x, y)
 
-  def roundTrip[A, L <: HList](m: A, t: BigQueryType[A] = BigQueryType[A])
+  implicit val userId = BigQueryMappableType.longBigQueryMappableType.xmap(UserId.apply)(_.id)
+
+  def roundTrip[A, L](m: A, t: BigQueryType[A] = BigQueryType[A])
```
Comment (nevillelyh) [inline bigquery/src/test/scala/shapeless/datatype/bigquery/BigQueryTypeSpec.scala]: nitpick: align implicit args?

File: core/src/test/scala/shapeless/datatype/record/RecordMatcherSpec.scala [hunk 2]
```diff
@@ -22,15 +22,14 @@ class RecordMatcherSpec extends Properties("RecordMatcher") {
   object inc extends ->((x: Double) => x + 1.0)
   implicit def compareDoubles(x: Double, y: Double) = math.abs(x) == math.abs(y)
 
-  def test[A, L <: HList](original: A, withNegate: A, withInc: A,
-                          t: RecordMatcher[A] = RecordMatcher[A])
+  def test[A, L](original: A, withNegate: A, withInc: A)
```
Comment (nevillelyh) [inline core/src/test/scala/shapeless/datatype/record/RecordMatcherSpec.scala]: nitpick: align implement args?

File: test/src/test/scala/shapeless/datatype/test/Records.scala [hunk 3]
```diff
@@ -5,22 +5,26 @@ import org.joda.time.Instant
...
 
 object Records {
+  case class UserId(id: Long)
+
+  case class User(userId: UserId, name: String)
+
   case class Required(booleanField: Boolean,
                       intField: Int, longField: Long, floatField: Float, doubleField: Double,
                       stringField: String,
-                      byteStringField: ByteString, byteArrayField: Array[Byte],
```
Comment (nevillelyh) [inline test/src/test/scala/shapeless/datatype/test/Records.scala]: Why removing these `byteStringField: List[ByteString]` here? Are you trying to decouple protobuf dependency?

Comment (Fristi) [inline test/src/test/scala/shapeless/datatype/test/Records.scala]: You could potentially! But I can bring back the `List[ByteString]`

Comment (nevillelyh) [inline test/src/test/scala/shapeless/datatype/test/Records.scala]: Yeah would be nice to bring them back.

**Comment by nevillelyh** [comment]
This is out of date, gonna close for now and try to work some changes into the code base.

**Pull request closed without merging**